portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,54 @@
 
 package com.google.sps.servlets;
 
+import com.google.gson.Gson;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private final Gson GSON = new Gson();
+  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/plain;");
-    response.getWriter().println("My name is Bao Xian!");
+    List<String> comments = new ArrayList<>();
+
+    Query query = new Query("Message");
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()){
+      comments.add((String) entity.getProperty("message"));
+    }
+
+    String json = GSON.toJson(comments);
+
+    response.setContentType("text/json;");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String message = request.getParameter("message");
```

Comment:
just wondering, but do you want empty string to be put as comment?

Reply:
Nope! Was planning to add the validation on the front-end later on, is that a good strategy?

Reply:
No, validation on frontend is an optional for better user experience, however validation on backend is a must.
the reason is because frontend code is modifiable (for example by using inspect)

Reply:
got it, added in!